text/0000-symbol-name-mangling-v2.md

```diff
@@ -0,0 +1,1156 @@
+
+<backref> = "B" <base-62-number>
+
+// We use <path> here, so that we don't have to add a special rule for
+// compression. In practice, only a crate root is expected.
+<instantiating-crate> = <path>
+```
+
+### Namespace Tags
+
+Namespaces are identified by an implementation defined single character tag
+(the `<ns>` production). Only closures (`C`) and shims (`S`) have a
+specific character assigned to them so that demanglers can reliable
+adjust their output accordingly. Other namespace tags have to be omitted
+or shown verbatim during demangling.
+
+This is a concession to the compiler's current implementation. While the
+language only knows two namespaces (the type and the value namespace), the
+compiler uses many more in some important data structures and disambiguation
+indices are assigned according to these internal data structures. So, in
+order not to force the compiler to waste processing time on re-constructing
+different disambiguation indices, the internal unspecified "namespaces" are
+used. This may change in the future.
+
+### Type-Level Constants
+
+As described above, the grammar encodes constant values via the
+`<const-data> = {<hex-digit>} "_"` production, where `{<hex-digit>}` is
+the numeric value of the constant, not its representation as bytes. Using
+the numeric value is platform independent but does not easily scale to
+non-integer data types.
+
+In the future it is likely that Rust will support complex type-level
+constants (i.e. not just integers). This RFC suggests to develop a
+proper mangling for these as part of the future const-generics work,
+and, for now, only define a mangling for integer values.
+
+### Punycode Identifiers
+
+Punycode generates strings of the form `([[:ascii:]]+-)?[[:alnum:]]+`.
+This is problematic for two reasons:
+
+- Generated strings can contain a `-` character; which is not in the
+  supported character set.
+- Generated strings can start with a digit; which makes them clash
+  with the byte-count prefix of the `<identifier>` production.
+
+For these reasons, vanilla Punycode string are further encoded during mangling:
+
+- The `-` character is simply replaced by a `_` character.
+- The part of the Punycode string that encodes the non-ASCII characters
+  is a base-36 number, using `[a-z0-9]` as its "digits". We want to get
+  rid of the decimal digits in there, so we simply remap `0-9` to `A-J`.
+
+Here are some examples:
+
+| Original        | Punycode        | Punycode + Encoding |
+|-----------------|-----------------|---------------------|
+| føø             | f-5gaa          | f_Fgaa              |
+| α_ω             | _-ylb7e         | __ylbHe             |
+| 铁锈             | n84amf          | nIEamf              |
+| 🤦              | fq9h            | fqJh                |
+| ρυστ            | 2xaedc          | Cxaedc              |
+
+With this post-processing in place the Punycode strings can be treated
+like regular identifiers and need no further special handling.
```

Comment:
Punycode feels like an extremely arbitrary choice, especially given that it is desirable to keep symbols short and that we could encode with up to base-52 (a density upgrade of up-to ~30%) using the whole range `a-zA-Z`.

I feel like the engineering effort to describe such encoding mechanism tailored for our use cases (even if it is ideologically similar to punycode) is worth the effort. Especially so, given that the encoding algorithm is not exactly punycode anymore after these modifications, which makes any compatibility or ease-of-use claims invalid.

Reply:
I have never seen a non-ASCII character in a Rust identifier in the wild, which suggests to me that minimizing their mangled representation isn't worth any effort.

Reply:
@rocallahan You likely haven't seen them because they're [not yet stable.](https://github.com/rust-lang/rust/issues/55467)

Reply:
Ah, good point!

I'm still skeptical that they'll become significant even if and when they are stable. They've been allowed in Java for many years but I don't recall ever seeing a Java API that uses them.

Reply:
Of course, doing so isn't stable, so most people who might use non-ascii idents but can be on stable otherwise would prefer latinizing the non-english idents. And of course, unless you speak a language that uses a non-latin script, the likelyhood of you interacting with a codebase that would use a non-ascii ident is relatively close to 0%. (But of course this was all hashed out multiple times in the ascii-idents rfc and does not need to be rehashed here.)

I agree that punycode is "good enough", especially as it's standard, but I disagree that a small amount of effort to specify "punycode but with a larger encoding space" would not be practically useful (thus worth it).

Reply:
The choice of Punycode is not arbitrary. The reasons are:
- It's standardized and probably better than anything we could come up with, given that we don't have an arbitrary amount of resources.
- It retains a good deal of readability for mixed ASCII/non-ASCII strings.
- There's precedent for using it in a mangling scheme (Swift does it).

The only real alternative that I'd be up for is something super simple, like a base62 encoding. That would give up readability for ease of implementation.